feat: delegated identity providers and accelerated IdP flow (#423)
Universal-Commerce-Protocol/ucp: docs/specification/identity-linking.md

Reviewer: This suggestion anchors discovery on the RFC 9728 PRM's `authorization_servers` instead of hard-coding issuer = business domain. The byte-match must target the `authorization_servers` entry and the RFC 8414 URL must use sect 3.1 path insertion.

The business's own AS lives in `authorization_servers`, external IdPs stay in `config.providers`. This separates OAuth discovery for the trust domain from the chaining discovery in UCP. 

This also addresses @amithanda's [concerns](https://github.com/Universal-Commerce-Protocol/ucp/pull/423#discussion_r3297173348) on the case when the authorization server is different from the business domain.

```suggestion
A business exposes two OAuth roles, which **MAY** live on different hosts: the
**protected resource** (its UCP API, identified by the business domain in the
UCP profile) and the **authorization server** (AS) that issues tokens for it.
Platforms discover the AS starting from the protected resource.

**Step 1 — Resolve the authorization server.** Platforms determine the AS from
the resource's protected resource metadata
([RFC 9728](https://datatracker.ietf.org/doc/html/rfc9728){ target="_blank" }):

* Fetch `{business-domain}/.well-known/oauth-protected-resource`, or use the
  `resource_metadata` URL from a `WWW-Authenticate` challenge the resource
  returned
  ([RFC 9728 §5.1](https://datatracker.ietf.org/doc/html/rfc9728#section-5.1){ target="_blank" }).
* The metadata's `resource` value **MUST** identify the business resource the
  platform is accessing; platforms **MUST** reject metadata whose `resource`
  does not match
  ([RFC 9728 §3.3](https://datatracker.ietf.org/doc/html/rfc9728#section-3.3){ target="_blank" }).
* The platform **MUST** select an issuer from the non-empty
  `authorization_servers` array and use it as the **AS issuer** for the rest of
  discovery. This issuer is authoritative and **MAY** be hosted on a different
  origin than the business domain.
* If the resource publishes no protected-resource metadata, the AS issuer
  defaults to the business domain (single-host deployments).

**Step 2 — Fetch authorization server metadata.** Using the AS issuer from
Step 1, resolve metadata via a strict two-tier hierarchy. URLs are constructed
per [RFC 8414 §3.1](https://datatracker.ietf.org/doc/html/rfc8414#section-3.1){ target="_blank" }:
when the issuer has a path component, the well-known segment is **inserted
between the host and the path**, not appended.

1. **RFC 8414 (Primary):** Fetch
   `https://{issuer-host}/.well-known/oauth-authorization-server{issuer-path}`.
    * `2xx` response: use this metadata. Discovery complete.
    * `404 Not Found`: proceed to step 2.
    * Any other non-2xx response, network error, or timeout: **MUST** abort.
      **MUST NOT** proceed to step 2.

2. **OIDC Discovery (Fallback):** Fetch
   `{issuer}/.well-known/openid-configuration`.
    * `2xx` response: use this metadata. Discovery complete.
    * Any non-2xx response, network error, or timeout: **MUST** abort the
      identity linking process.

Platforms **MUST NOT** silently fall through on any error other than `404` in
step 1. This prevents partial or undefined behavior when a server is
misconfigured or temporarily unavailable.

**Step 3 — Validate.** The `issuer` value in the discovered metadata **MUST**
match the AS issuer selected in Step 1 exactly, byte-for-byte (per
[RFC 8414 §3.3](https://datatracker.ietf.org/doc/html/rfc8414#section-3.3){ target="_blank" }).
Platforms **MUST NOT** normalize (e.g., strip trailing slashes) before
comparison. The match target is the selected `authorization_servers` entry —
not the business domain.
```

Author: @shopdave 👍  on the aim but feels like we're replicating a good chunk of normative language defined by the RFCs we're citing. Attempt at a compressed form: https://github.com/Universal-Commerce-Protocol/ucp/pull/423/changes/0932c8e27d3be1d868a9cb3aec88891efdeaab2a -- does that look reasonable?

Reviewer: LGTM!

Reviewer: We have now arrived at this future. :-) 

```suggestion
    domain conventions. The business **MUST** publish this metadata when
    the authorization server does not live on the business domain.
```

where `code_verifier` is absent or does not verify against the stored
`code_challenge`.

## Identity Providers

The `config.providers` map declares the trusted identity providers a
business accepts for authentication.

> * **When absent:** by advertising the `dev.ucp.common.identity_linking`

Reviewer 2: Is this for backwards compatibility? Why have this implicit behavior?

Reviewer: > Is this for backwards compatibility? Why have this implicit behavior?

Correct, the idea is to retain the same semantics for the regular identity linking flow as defined in the very first version of the specification.

Author: Updated the contract: providers is for delegated allowlist. This clause is no longer necessary.

    capability without a `providers` map, the business implicitly declares
    itself as its own (and only) identity provider. Platforms run OAuth 2.0
    directly against the business domain using the [Discovery](#discovery)
    hierarchy.
* **When present:** the map is a strict whitelist. Platforms **MUST**
    select from the listed providers and **MUST NOT** fall back to direct
    OAuth against the business domain. To accept direct OAuth alongside
    external IdPs, a business **MUST** self-list (own `auth_url` resolving
    to own authorization server); omitting self signals IdP-mediated
    authentication only.

When a platform already holds a token from a listed provider (e.g., from a
prior account-linking flow with that IdP), it can chain the user's identity
to the business via the [Accelerated IdP Flow](#accelerated-idp-flow) — no
fresh browser-based OAuth flow required.

### Provider Configuration

Each provider entry specifies the discovery base URL:

| Field | Type | Required | Description |
| :---- | :--- | :------- | :---------- |

> | `type` | string | No (default: `oauth2`) | Provider type discriminator. |

Reviewer 2: Again, why a default here?

Reviewer: > Again, why a default here?

Purely convenience. Want to make clear that unless otherwise noted, OAuth is the protocol for identity linking.

| `auth_url` | string (URI) | Yes | Base URL for authorization server metadata discovery. |

Platforms **MUST** discover the provider's authorization server metadata from
`auth_url` using the same two-tier hierarchy as [Discovery](#discovery)
(RFC 8414 primary, OIDC fallback on 404 only).

> ### Provider Selection

Reviewer 3: Should we introduce an explicit differentiator (such as a reserved provider key `"self"` or a boolean flag `"self": true` / `"direct": true`) in the provider schema to help platforms distinguish between self-listed direct OAuth endpoints and external delegated IdPs?

It could be argued that platforms can identify the direct endpoint by matching the storefront domain with the provider's `auth_url`, or by treating all listed providers uniformly (running Token Exchange for every provider). However, both the direct OAuth fallback and the proposed provider schema together are insufficient:
1. **Strict Issuer Validation Forces Self-Listing:**
   UCP mandates that: *\"The `issuer` value in the discovered metadata **MUST** match the discovery base URI exactly (per RFC 8414 §3.3). Platforms **MUST NOT** normalize (e.g., strip trailing slashes) before comparison — the value must be a byte-for-byte match.\"*
   If a merchant separates storefront hosting from customer identity (e.g., storefront on `brand.com` but AS on Okta at `brand.okta.com`), direct OAuth discovery on the storefront domain will fail this strict issuer check. To bypass this domain mismatch, the merchant is forced to self-list their AS domain in the `providers` map.
2. **Structural Ambiguity and Circular Requests:**
   Once self-listed, the merchant's endpoint is structurally identical to external IdPs (such as Google). Because the domains do not match the storefront, platforms cannot use domain-matching heuristics to identify the direct endpoint. Without a clear differentiator, the platform cannot determine which endpoint requires direct redirects and which enables back-channel token exchange. If the platform treats the self-listed endpoint as an external IdP, it will attempt a redundant, circular token exchange (RFC 8693) on the merchant's own AS, causing latency and request failures.
An explicit flag resolves this structural ambiguity and ensures platforms can execute direct OAuth redirects for self-hosted authorization and reserve token exchange logic solely for external federators.

Reviewer: I believe this is a valid concern, and I have proposed changes to the Discovery section. The missing part here is how authorization servers get discovered from the resource server.

> ### Provider Selection

Reviewer 3: Should we introduce an explicit differentiator (such as a reserved provider key `"self"` or a boolean flag `"self": true` / `"direct": true`) in the provider schema to help platforms distinguish between self-listed direct OAuth endpoints and external delegated IdPs?

It could be argued that platforms can identify the direct endpoint by matching the storefront domain with the provider's `auth_url`, or by treating all listed providers uniformly (running Token Exchange for every provider). However, both the direct OAuth fallback and the proposed provider schema together are insufficient:
1. **Strict Issuer Validation Forces Self-Listing:**
   UCP mandates that: *\"The `issuer` value in the discovered metadata **MUST** match the discovery base URI exactly (per RFC 8414 §3.3). Platforms **MUST NOT** normalize (e.g., strip trailing slashes) before comparison — the value must be a byte-for-byte match.\"*
   If a merchant separates storefront hosting from customer identity (e.g., storefront on `brand.com` but AS on Okta at `brand.okta.com`), direct OAuth discovery on the storefront domain will fail this strict issuer check. To bypass this domain mismatch, the merchant is forced to self-list their AS domain in the `providers` map.
2. **Structural Ambiguity and Circular Requests:**
   Once self-listed, the merchant's endpoint is structurally identical to external IdPs (such as Google). Because the domains do not match the storefront, platforms cannot use domain-matching heuristics to identify the direct endpoint. Without a clear differentiator, the platform cannot determine which endpoint requires direct redirects and which enables back-channel token exchange. If the platform treats the self-listed endpoint as an external IdP, it will attempt a redundant, circular token exchange (RFC 8693) on the merchant's own AS, causing latency and request failures.
An explicit flag resolves this structural ambiguity and ensures platforms can execute direct OAuth redirects for self-hosted authorization and reserve token exchange logic solely for external federators.

Author: Great catch. I think the cleaner approach is to flip the contract: providers is the delegated+approved set. PTAL at the new algorithm.

to the business's token endpoint. The business validates the grant and issues
its own access token under its own authority.

This flow profiles the identity and authorization chaining pattern in
[draft-ietf-oauth-identity-chaining](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-identity-chaining-08){ target="_blank" }.
UCP tightens two aspects beyond what the base RFCs mandate: token exchange
**MUST** use `resource` (not `audience`), and the JWT authorization grant
**MUST** carry `aud` as a single-valued URI plus a unique `jti` (see
[JWT Authorization Grant](#jwt-authorization-grant)).

### Flow

1. Platform discovers `config.providers` in the business's identity linking
   capability and selects a provider it already holds a valid token for.
2. Platform requests a JWT authorization grant from the IdP via token
   exchange ([RFC 8693](https://datatracker.ietf.org/doc/html/rfc8693){ target="_blank" })
   at the IdP's token endpoint:
    * `grant_type`: `urn:ietf:params:oauth:grant-type:token-exchange`
    * `subject_token`: the platform's existing IdP access token
    * `subject_token_type`: `urn:ietf:params:oauth:token-type:access_token`

>     * `resource`: the business's authorization server issuer URI (the IdP

Reviewer 3: Should we relax the strict requirement to use the `resource` parameter and instead permit the use of `audience`, `resource`, or both (containing the business AS issuer URI or logical identifier) during token exchange?
**Why:**
Strictly mandating `resource` and prohibiting `audience` (lines 433-436) introduces a critical interoperability barrier with several widely deployed Identity Providers (IdPs) and Security Token Services (STS):
1. **Google Cloud STS:** Google’s Workload Identity Federation STS endpoint strictly requires the `audience` parameter (mapped to the workload pool provider URI) and does not support or recognize the `resource` parameter.
2. **Keycloak:** In Keycloak's RFC 8693 token exchange implementation, downscoping and client targets are resolved via the `audience` parameter (mapping to Keycloak client IDs). Keycloak historically does not natively resolve the `resource` parameter out-of-the-box.

**RFC 8693 Semantics:**
While it may be argued that `resource` is defined in RFC 8693 as a URI and `audience` as a logical name, Section 2.1 of the RFC explicitly states that the `audience` value *\"MAY be a URI\"*. Therefore, passing the merchant's AS issuer URI in the `audience` parameter is fully compliant with the RFC and ensures out-of-the-box interoperability with standard OIDC and STS implementations without compromising security.

with two UCP-specific constraints:

* `aud` **MUST** be a single value (the business's AS issuer URI), not an
  array — chaining targets one business per grant.
* `jti` **MUST** be present (RFC 7523 says MAY) so businesses can enforce
  single-use replay protection (see [Security Considerations](#security-considerations)).

`exp` **SHOULD** be no more than 60 seconds after `iat`. The IdP **MAY**
include additional claims to convey authorization context (consent records,
user attributes, etc.).

### Business Token Issuance

Upon receiving a JWT authorization grant at its token endpoint, the
business's authorization server **MUST** validate the assertion per
[RFC 7523 §3](https://datatracker.ietf.org/doc/html/rfc7523#section-3){ target="_blank" },
with the following UCP-specific requirements:

* `iss` **MUST** identify a provider listed in `config.providers`.
* `aud` **MUST** match the business's own AS issuer URI exactly.
* The JWT signature **MUST** be verified against the IdP's `jwks_uri`; if
  JWKS cannot be retrieved, the business **MUST** fail closed.

Once the assertion is validated, the business resolves the user from `sub`

> (auto-provisioning permitted) and issues an access token scoped to the

Reviewer 4: Worth _considering_ whether some guidance on identity claims would help implementers here — a couple of scenarios where opaque `sub` could be tricky:
                                                                                                                               
1. Cross-IdP user resolution. A business that accepts both com.shopify.accounts and com.google receives sub=abc123 from one and sub=xyz789 from the other. If these correspond to the same end user, the business has no way to correlate them without a shared identifier like email. Without that, a user who links via Google and later links via the business oauth could end up with two disjoint accounts at the same business.   This has happened to me a time or two in fact ;-)
                          
2. Auto-provisioning a new account. A business receives a chained grant for a sub it has never seen. To create a useful account it likely needs at minimum a contact identifier (email, phone) — otherwise it provisions an empty record it can't communicate with or match to an existing customer.
                                                                                                                               
[OIDC Core §5.1](https://openid.net/specs/openid-connect-core-1_0-31.html#StandardClaims) defines standard claims (email, email_verified, name, etc.) that seem like a natural fit here. Not necessarily as a MUST, but perhaps a SHOULD for IdPs to include email and email_verified when available, or a note in the spec acknowledging that businesses will likely need claims beyond sub for practical user resolution. The building blocks already exist via OIDC — it might just be a matter of connecting the reference.

Author: Good catch, updated.

> requested UCP scopes. If user interaction is required (terms acceptance,

Reviewer 4: "If user interaction is required... " should UCP require businesses to issue all advertised scopes _regardless_ of which identity provider is used? e.g. would a business have any cause to require use of _it's own_ OAuth endpoints to acquire certain scopes? ... I don't have a specific scenario in mind but possibly useful to cover this expectation in the docs?

Author: Clarified.

where `code_verifier` is absent or does not verify against the stored
`code_challenge`.

## Identity Providers

The `config.providers` map declares external trusted identity providers
from which the business will accept chained identity via JWT bearer
assertions for the [Accelerated IdP Flow](#accelerated-idp-flow). It is
additive metadata on top of the always-available direct OAuth path
against the business domain (see [Discovery](#discovery)) — never a

> closed whitelist.

Reviewer: Nit: The text calls `providers` "never a closed whitelist," but for chaining it is one. Lines around 511 requires `iss` to be a listed provider. Suggest distinguishing the two paths explicitly.

```suggestion
against the business domain (see [Discovery](#discovery)); for the chaining
path, it is a closed allowlist — a business **MUST** reject a JWT
authorization grant whose `iss` is not a listed provider (see
[Business Token Issuance](#business-token-issuance)).
```

    degenerate (the same server would issue and validate the assertion),
    and direct OAuth is already available via [Discovery](#discovery).
    Platforms **MUST** ignore any entry whose `auth_url` matches the
    business's own issuer URI.

### Provider Configuration

Each provider entry is keyed by a reverse-domain identifier and described by
its `type`:

| Field | Type | Required | Description |
| :---- | :--- | :------- | :---------- |
| `type` | string | Yes | Provider mechanism discriminator. `oauth2` is the only type defined in this version; future versions **MAY** define additional types as non-breaking extensions. |
| `auth_url` | string (URI) | Yes, for `oauth2` | Base URL for authorization server metadata discovery. |

The `type` value is an open string, not a closed enumeration. Platforms
**MUST** treat provider entries whose `type` they do not support as filtered
out (see [Provider Selection](#provider-selection)) rather than rejecting the
business's configuration.

For `oauth2` providers, platforms **MUST** discover the authorization server
metadata from `auth_url` using the same two-tier hierarchy as

> [Discovery](#discovery) (RFC 8414 primary, OIDC fallback on 404 only).

Reviewer: Align with the proposed changes to Discovery.

```suggestion
metadata from `auth_url` using the same two-tier metadata hierarchy as
[Discovery](#discovery) Step 2 (RFC 8414 primary with §3.1 path insertion,
OIDC fallback on 404 only), treating `auth_url` as the issuer. The
protected-resource step does not apply — `auth_url` is already the IdP
issuer — and the business validates the JWT grant's `iss` against it per
[Business Token Issuance](#business-token-issuance).
```

* **Token revocation.** Platforms **MUST** revoke user identity tokens at the
  business's revocation endpoint (RFC 7009) when a user unlinks their account.
  Businesses **MUST** reject subsequent requests that present revoked tokens.
* **JWT grant lifetime.** JWT authorization grants **MUST** be short-lived;
  the `exp` claim **SHOULD** be no more than 60 seconds after `iat`. Short
  lifetimes limit the window for grant theft and replay.

> * **JWT grant single-use.** Businesses **SHOULD** enforce single-use JWT

Reviewer: We already mandate `jti` presence, so we might as well make enforcement mandatory.

```suggestion
* **JWT grant single-use.** Businesses **MUST** enforce single-use JWT;
  a short exp narrows the replay window, but only jti tracking closes it.
```

  authorization grants by tracking the `jti` claim within the grant's
  validity window.
* **Grant relay.** Businesses **MUST NOT** store or forward JWT
  authorization grants received from platforms. Grants are bearer
  credentials scoped to a single audience (`aud`) and a single use.
* **Cross-IdP account linking.** Federation collapses trust boundaries:
  a business that auto-links accounts across listed providers based on
  any IdP-asserted claim (email, phone, name) extends each provider's
  verification process into account-takeover risk. A provider that
  issues `email_verified=true` for an email the user does not control
  can hijack any account at the business sharing that email. The stable
  per-IdP identifier is `(iss, sub)`; other claims are advisory.

>   Businesses **SHOULD** require user-mediated linking — the user

Reviewer 4: 👍

* **Token revocation.** Platforms **MUST** revoke user identity tokens at the
  business's revocation endpoint (RFC 7009) when a user unlinks their account.
  Businesses **MUST** reject subsequent requests that present revoked tokens.
* **JWT grant lifetime.** JWT authorization grants **MUST** be short-lived;
  the `exp` claim **SHOULD** be no more than 60 seconds after `iat`. Short
  lifetimes limit the window for grant theft and replay.

> * **JWT grant single-use.** Businesses **MUST** enforce single-use JWT;

Reviewer 3: I think the "JWT grant single-use" bullet is broken mid-clause:

> Businesses **MUST** enforce single-use JWT; a short exp narrows the replay window, but only jti tracking closes it. authorization grants by tracking the `jti` claim within the grant's validity window.

This is probably a merge artifact from @shopdave's suggestion. I think, two fragments got spliced. 

```suggestion
* **JWT grant single-use.** Businesses **MUST** enforce single-use of JWT
  authorization grants by tracking the `jti` claim within the grant's
  validity window. A short `exp` narrows the replay window, but only
  `jti` tracking closes it.
```
